src/Sema.zig

```diff
@@ -23188,35 +23191,40 @@ fn coerceExtra(
                     },
                     else => {},
                 },
-                .Slice => {
-                    // pointer to tuple to slice
-                    if (inst_ty.isSinglePointer() and inst_ty.childType().isTuple() and dest_info.size == .Slice and
-                        sema.checkPtrAttributes(dest_ty, inst_ty, &in_memory_result))
-                    {
-                        return sema.coerceTupleToSlicePtrs(block, dest_ty, dest_ty_src, inst, inst_src);
+                .Slice => to_slice: {
+                    if (inst_ty.zigTypeTag() == .Array) {
+                        return sema.fail(
+                            block,
+                            inst_src,
+                            "array literal requires address-of operator (&) to coerce to slice type '{}'",
+                            .{dest_ty.fmt(sema.mod)},
+                        );
                     }
 
+                    if (!inst_ty.isSinglePointer()) break :to_slice;
+                    const inst_child_ty = inst_ty.childType();
+                    if (!inst_child_ty.isTuple()) break :to_slice;
+
                     // empty tuple to zero-length slice
                     // note that this allows coercing to a mutable slice.
-                    if (inst_ty.isSinglePointer() and
-                        inst_ty.childType().tag() == .empty_struct_literal and
-                        dest_info.size == .Slice)
-                    {
+                    if (inst_child_ty.tupleFields().types.len == 0) {
                         const slice_val = try Value.Tag.slice.create(sema.arena, .{
                             .ptr = Value.undef,
                             .len = Value.zero,
                         });
                         return sema.addConstant(dest_ty, slice_val);
                     }
 
-                    if (inst_ty.zigTypeTag() == .Array) {
+                    // pointer to tuple to slice
+                    if (dest_info.mutable) {
                         return sema.fail(
                             block,
                             inst_src,
-                            "array literal requires address-of operator (&) to coerce to slice type '{}'",
+                            "cannot cast pointer to tuple to '{}'",
```

Comment:
It'd be nice if this explained why it cannot do the cast.

Reply:
How does something like
```
coercion from tuple pointer to {:0} would create constant temporary, but {:0} can only point to mutable memory.
```
sound?

Reply:
Well, it's true but also quite hard to understand. I was thinking of something more in the lines of `note: pointers to tuples can only coerce to constant pointers`.

Reply:
Sorry to inject thrash here, but I think the message needs to be specific to array/slice pointers, since `@as(*std.meta.Tuple(&.{usize, usize}), &.{x, x + 1})` is perfectly legal

How about `note: Tuple-to-array conversion requires a destination type of [N]T, *const [N]T, or []const T.` ? Not terribly short, but I haven't been able to come up with better. Or perhaps `note: a tuple can only coerce to an array type directly, or to a const pointer/slice of an array.`